sol/main.c

```diff
@@ -0,0 +1,459 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <rte_sched.h>
+
+#include "gatekeeper_gk.h"
+#include "gatekeeper_launch.h"
+#include "gatekeeper_sol.h"
+
+/*
+ * Gatekeeper request priority queue implementation.
+ *
+ * To implement the request PQ, we maintain a linked list of packets
+ * listed in order of highest priority to lowest priority. We keep an
+ * array where each index represents a priority, and each element of
+ * that array holds a reference to the last packet of that priority.
+ * This allows us to quickly insert new packets of any priority into
+ * the linked list and drop the packet of lowest priority if necessary.
+ */
+
+/* The request priority queue is allocated 5% of the link's bandwidth. */
+#define REQ_QUEUE_BW_RATE (.05)
+
+/* Token bucket rate approximation error. */
+#define GATEKEEPER_TB_RATE_CONFIG_ERR (1e-7)
+
+/*
+ * Scaling for cycles_per_byte calculation; chosen so that minimum rate
+ * is 480 bit/sec. Adapted for use from the DPDK scheduling library.
+ */
+#define GATEKEEPER_TIME_SHIFT (8)
+
+/*
+ * Solicitor configuration allocated by alloc_sol_conf(), needed
+ * by gk_solicitor_enqueue() since the GK instances do not have
+ * any visibility into the Solicitor.
+ */
+static struct sol_config *sol_conf;
+
+static void
+insert_new_priority_req(struct req_queue *req_queue, struct priority_req *pr,
+	uint8_t priority)
+{
+	uint8_t next;
+
+	req_queue->priorities[priority] = pr;
+
+	/* This is the first packet in the queue. */
+	if (req_queue->len == 0) {
+		list_add(&pr->list, &req_queue->head);
+		req_queue->highest_priority = priority;
+		req_queue->lowest_priority = priority;
+		return;
+	}
+
+	/* Not the first packet, but still insert at the head of the queue. */
+	if (priority > req_queue->highest_priority) {
+		list_add(&pr->list, &req_queue->head);
+		req_queue->highest_priority = priority;
+		return;
+	}
+
+	/* Insert in middle or end of queue. */
+	if (priority < req_queue->lowest_priority)
+		req_queue->lowest_priority = priority;
+	for (next = priority + 1; next < req_queue->highest_priority; next++) {
+		if (req_queue->priorities[next] != NULL) {
+			list_add(&pr->list, &req_queue->priorities[next]->list);
+			return;
```

Comment:
It can't be the case that `priority == req_queue->highest_priority` because this function only inserts a request for a priority that does not currently exist in the queue. But I will add an `RTE_VERIFY()` to make sure this expectation is met.